core: crypto: add support for SM2 PKE

jforissier opened the pull request:
Some more Chinese crypto code, this time SM2 public key cryptography (based in elliptic curves).
As mentioned in the main commit, SM2 also defines a signature algorithm and a key exchange protocol. I plan to add them next.

Note: the first two commits are related to LibTomCrypt, therefore I have opened separate PRs against the LTC import branch.

jenswi-linaro commented on core/lib/libtomcrypt/sm2.c:
```diff
@@ -0,0 +1,632 @@
+ * Encryption algorithm
+ */
+TEE_Result crypto_acipher_sm2_pke_encrypt(struct ecc_public_key *key,
+					  const uint8_t *src __unused,
```
`__unused` should be removed here and for `dst` below.

jenswi-linaro commented on core/lib/libtomcrypt/sm2.c:
```diff
@@ -0,0 +1,632 @@
+		goto out;
+	res = crypto_hash_final(ctx, u, sizeof(u));
+	if (res)
+		goto out;
```
We should probably use the LTC internal hash interface instead, or this can't be upstreamed.

jforissier replied:
Problem is, the hash is SM3 which is not supported by LTC.

jenswi-linaro replied:
OK

jforissier commented:
> Comments for the commit "core: crypto: add support for SM2 PKE"
> 
> I'd prefer if this commit was split in two commits. One for the LTC changes (with updated to the import branch) and another for the rest.

OK. I wasn't 100% sure what should go into the import branch, I mean: is it everything commit that touches `libtomcrypt/*` or just `libtomcrypt/{include,src}`. `libtomcrypt/*.c` are essentially interface with OP-TEE core and additional algorithms (like this PR). Now I think it is best to have all changes to `libtomcrypt/*` be made in the import branch.

I will split the code and create a separate PR.

> Are we planning to upstream any of this?

I considered this option, but it seemed like significantly more work: SM2 needs the SM3 hash, so SM3 should be upstream too; I am not super comfortable with the LTC interfaces, coding style, test cases... Also, by keeping SM2 as proposed, we have a template ready to use in case someone wants to add support for Mbed TLS as the underlying ECC provider.

jforissier commented:
I have split "core: crypto: add support for SM2 PKE" and created a PR against the LTC import branch: https://github.com/OP-TEE/optee_os/pull/3503 (note:I have renamed `sm2.c` to `sm2-pke.c`).

I will close this PR and create a v2 for the remaining code.

jforissier commented:
Superseded by https://github.com/OP-TEE/optee_os/pull/3504.
